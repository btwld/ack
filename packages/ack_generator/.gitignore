# AI Agent Instructions (local only)
.augment-guidelines
.ai/
CLAUDE.md

# Dart/Flutter
.dart_tool/
.packages
build/
