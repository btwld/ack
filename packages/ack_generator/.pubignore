# Development and IDE files
melos_ack_generator.iml
pubspec_overrides.yaml

# Development documentation
docs/

# Local example (if any development examples exist)
example/

# Tool scripts
tool/
