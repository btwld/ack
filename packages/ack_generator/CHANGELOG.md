## 1.0.0-beta.1 (2025-10-06)

* See [release notes](https://github.com/btwld/ack/releases/tag/v1.0.0-beta.1) for details.
