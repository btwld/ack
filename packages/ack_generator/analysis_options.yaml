include: package:lints/recommended.yaml

analyzer:
  exclude:
    - test/fixtures/**
    - "**/*.g.dart"

linter:
  rules:
    - require_trailing_commas
    - prefer_single_quotes
