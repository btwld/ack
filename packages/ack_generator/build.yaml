builders:
  ack_generator:
    import: "package:ack_generator/builder.dart"
    builder_factories: ["ackGenerator"]
    build_extensions: {".dart": [".g.dart"]}
    auto_apply: dependents
    build_to: source
    required_inputs: [".dart"]

# Global configuration for performance
global_options:
  ack_generator|ack_generator:
    options:
      # Enable verbose logging for debugging
      verbose: false
