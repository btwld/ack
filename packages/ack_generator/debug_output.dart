import 'package:ack_generator/src/builders/schema_builder.dart';
import 'package:ack_generator/src/models/model_info.dart';
import 'package:ack_generator/src/models/field_info.dart';

void main() {
  final builder = SchemaBuilder();
  
  final model = ModelInfo(
    className: 'User',
    schemaClassName: 'UserSchema', 
    description: 'User model for testing',
    fields: [
      FieldInfo(
        name: 'id',
        jsonKey: 'id',
        type: 'String',
        isRequired: true,
        isNullable: false,
        constraints: [],
        enumValues: [],
      ),
      FieldInfo(
        name: 'name', 
        jsonKey: 'name',
        type: 'String',
        isRequired: true,
        isNullable: false,
        constraints: [],
        enumValues: [],
      ),
    ],
    additionalProperties: false,
  );

  final result = builder.build(model);
  print('Generated output:');
  print('=' * 50);
  print(result);
  print('=' * 50);
}