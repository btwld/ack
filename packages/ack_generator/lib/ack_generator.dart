// AckGenerator provides annotations and code generation for validation schemas
export 'src/schema_model_generator.dart'
    hide
        PropertyInfo,
        PropertyConstraintInfo,
        RequiredConstraint,
        NullableConstraint,
        TypeName;
