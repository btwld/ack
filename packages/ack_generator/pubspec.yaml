name: ack_generator
description: Code generator for Ack schema validation models
version: 0.3.0-alpha.0
repository: https://github.com/btwld/ack
publish_to: none

environment:
  sdk: ^3.3.0

dependencies:
  # Core code generation dependencies
  analyzer: ^7.0.0
  build: ^2.4.0
  source_gen: ^2.0.0
  code_builder: ^4.10.0
  
  # Our packages
  ack:
    path: ../ack
  ack_annotations:
    path: ../ack_annotations
  
  # Utilities
  collection: ^1.18.0
  path: ^1.9.0
  meta: ^1.11.0
  dart_style: ^3.1.0

dev_dependencies:
  build_runner: ^2.4.0
  build_test: ^3.1.0
  test: ^1.25.0
  
  # Code quality
  lints: ^5.0.0
  very_good_analysis: ^9.0.0
