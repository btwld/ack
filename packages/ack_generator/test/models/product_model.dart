import 'package:ack_generator/ack_generator.dart';

/// A product with validation
@Schema(
  description: 'A product model with validation',
  additionalProperties: true,
  additionalPropertiesField: 'metadata',
)
class Product {
  /// Unique identifier for the product
  @IsNotEmpty()
  final String id;

  /// Name of the product (cannot be empty)
  @IsNotEmpty()
  final String name;

  /// Product description
  @IsNotEmpty()
  final String description;

  /// Product price (should be positive)
  final double price;

  /// Optional URL to product image
  @Nullable()
  final String? imageUrl;

  /// Product category
  @Required()
  final Category category;

  /// Additional metadata for the product
  final Map<String, dynamic> metadata;

  /// Constructor for creating a product
  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    this.imageUrl,
    required this.category,
    this.metadata = const {},
  });
}

/// Category for organizing products
@Schema(
  description: 'A category for organizing products',
  additionalProperties: true,
  additionalPropertiesField: 'metadata',
)
class Category {
  /// Unique identifier for the category
  @IsNotEmpty()
  final String id;

  /// Name of the category (cannot be empty)
  @IsNotEmpty()
  final String name;

  /// Optional description of the category
  @Nullable()
  final String? description;

  /// Additional metadata for the category
  final Map<String, dynamic> metadata;

  /// Constructor for creating a category
  Category({
    required this.id,
    required this.name,
    this.description,
    this.metadata = const {},
  });
}
