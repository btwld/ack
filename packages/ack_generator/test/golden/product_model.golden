// GENERATED CODE - DO NOT MODIFY BY HAND

// ignore_for_file: unnecessary_this, unnecessary_new, unnecessary_const, prefer_collection_literals
// ignore_for_file: lines_longer_than_80_chars, unnecessary_null_checks, non_constant_identifier_names

part of 'product_model.dart';

// **************************************************************************
// AckSchemaGenerator
// **************************************************************************

/// Generated schema for Product
/// A product model with validation
class ProductSchema extends SchemaModel<Product> {
  ProductSchema([Object? value = null]) : super(value);

  static final ObjectSchema schema = Ack.object(
    {
      'id': Ack.string.notEmpty(),
      'name': Ack.string.notEmpty(),
      'description': Ack.string.notEmpty(),
      'price': Ack.double,
      'imageUrl': Ack.string.nullable(),
      'category': CategorySchema.schema,
    },
    required: ['id', 'name', 'description', 'price', 'category'],
    additionalProperties: true,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<Product, ProductSchema>(
      (data) => ProductSchema(data),
    );
    // Register schema dependencies
    CategorySchema.ensureInitialize();
  }

  @override
  AckSchema getSchema() => schema;

  String get id => getValue<String>('id')!;

  String get name => getValue<String>('name')!;

  String get description => getValue<String>('description')!;

  double get price => getValue<double>('price')!;

  String? get imageUrl => getValue<String>('imageUrl');

  CategorySchema get category {
    return CategorySchema(getValue<Map<String, dynamic>>('category')!);
  }

  Map<String, Object?> get metadata {
    final map = toMap();
    final knownFields = {
      'id',
      'name',
      'description',
      'price',
      'imageUrl',
      'category'
    };
    return Map.fromEntries(
        map.entries.where((e) => !knownFields.contains(e.key)));
  }

  @override
  Product toModel() {
    if (!isValid) {
      throw AckException(getErrors()!);
    }
    return Product(
      id: id,
      name: name,
      description: description,
      price: price,
      imageUrl: imageUrl,
      category: category.toModel(),
      metadata: metadata,
    );
  }

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for Category
/// A category for organizing products
class CategorySchema extends SchemaModel<Category> {
  CategorySchema([Object? value = null]) : super(value);

  static final ObjectSchema schema = Ack.object(
    {
      'id': Ack.string.notEmpty(),
      'name': Ack.string.notEmpty(),
      'description': Ack.string.nullable(),
    },
    required: ['id', 'name'],
    additionalProperties: true,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<Category, CategorySchema>(
      (data) => CategorySchema(data),
    );
  }

  @override
  AckSchema getSchema() => schema;

  String get id => getValue<String>('id')!;

  String get name => getValue<String>('name')!;

  String? get description => getValue<String>('description');

  Map<String, Object?> get metadata {
    final map = toMap();
    final knownFields = {'id', 'name', 'description'};
    return Map.fromEntries(
        map.entries.where((e) => !knownFields.contains(e.key)));
  }

  @override
  Category toModel() {
    if (!isValid) {
      throw AckException(getErrors()!);
    }
    return Category(
      id: id,
      name: name,
      description: description,
      metadata: metadata,
    );
  }

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}
